File: apple/AITerminalCore/Sources/AITerminalCore/SSHConfigParser.swift
import Foundation

/// 解析 OpenSSH `~/.ssh/config`，把每个具体 Host 块转成 Connection。
/// 支持 Host / HostName / Port / User / IdentityFile；忽略通配 Host（含 * ?）。
public enum SSHConfigParser {

    private struct Block {
        var alias: String
        var hostName: String?
        var port: Int?
        var user: String?
        var identity: String?
    }

    public static func parse(_ text: String) -> [Connection] {
        var result: [Connection] = []
        var current: Block?

        func flush() {
            guard let b = current else { return }
            let host = (b.hostName?.isEmpty == false) ? b.hostName! : b.alias
            let auth: AuthType = (b.identity != nil) ? .privateKey : .password
            var conn = Connection(
                name: b.alias,
                host: host,
                port: b.port ?? 22,
                username: b.user ?? "",
                authType: auth
            )
            if let id = b.identity { conn.privateKeyPath = expandTilde(id) }
            result.append(conn)
        }

        for rawLine in text.split(whereSeparator: { $0 == "\n" || $0 == "\r" }) {
            let line = rawLine.trimmingCharacters(in: .whitespaces)
            if line.isEmpty || line.hasPrefix("#") { continue }

            // 拆成 key 与 value（分隔符：空白或 =）
            guard let splitIndex = line.firstIndex(where: { $0 == " " || $0 == "\t" || $0 == "=" }) else { continue }
            let key = line[..<splitIndex].lowercased()
            var value = line[line.index(after: splitIndex)...].trimmingCharacters(in: CharacterSet(charactersIn: " \t="))
            value = stripQuotes(value)

            switch key {
            case "host":
                flush()
                let firstPattern = value.split(separator: " ").first.map(String.init) ?? value
                if firstPattern.isEmpty || firstPattern.contains("*") || firstPattern.contains("?") {
                    current = nil
                } else {
                    current = Block(alias: firstPattern)
                }
            case "hostname":
                current?.hostName = value
            case "port":
                current?.port = Int(value)
            case "user":
                current?.user = value
            case "identityfile":
                if current?.identity == nil { current?.identity = value }
            default:
                break
            }
        }
        flush()
        return result.filter { !$0.host.isEmpty }
    }

    static func expandTilde(_ path: String) -> String {
        guard path.hasPrefix("~") else { return path }
        let home = FileManager.default.homeDirectoryForCurrentUser.path
        return home + path.dropFirst()
    }

    static func stripQuotes(_ s: String) -> String {
        if s.count >= 2, s.hasPrefix("\""), s.hasSuffix("\"") {
            return String(s.dropFirst().dropLast())
        }
        return s
    }
}

File: android/app/src/main/java/com/termind/app/SshConfigParser.kt
package com.termind.app

/** 解析 OpenSSH ~/.ssh/config 文本，把每个具体 Host 块转成 ServerConn（对齐 apple SSHConfigParser）。
 * 支持 Host / HostName / Port / User；忽略通配 Host（含 * ?）。私钥路径移动端无意义，按密码认证导入。 */
object SshConfigParser {

    private data class Block(var alias: String, var hostName: String? = null, var port: Int? = null, var user: String? = null)

    fun parse(text: String): List<ServerConn> {
        val result = mutableListOf<ServerConn>()
        var current: Block? = null

        fun flush() {
            val b = current ?: return
            val host = if (!b.hostName.isNullOrEmpty()) b.hostName!! else b.alias
            if (host.isEmpty()) return
            result.add(ServerConn(name = b.alias, host = host, user = b.user ?: "", port = b.port ?: 22))
        }

        for (rawLine in text.split('\n', '\r')) {
            val line = rawLine.trim()
            if (line.isEmpty() || line.startsWith("#")) continue
            // 拆 key/value（分隔符 空白或 =）
            val sepIdx = line.indexOfFirst { it == ' ' || it == '\t' || it == '=' }
            if (sepIdx < 0) continue
            val key = line.substring(0, sepIdx).lowercase()
            var value = line.substring(sepIdx + 1).trim().trim(' ', '\t', '=')
            if (value.length >= 2 && value.startsWith("\"") && value.endsWith("\"")) value = value.substring(1, value.length - 1)

            when (key) {
                "host" -> {
                    flush()
                    val firstPattern = value.split(" ").firstOrNull() ?: value
                    current = if (firstPattern.isEmpty() || firstPattern.contains("*") || firstPattern.contains("?")) null
                              else Block(alias = firstPattern)
                }
                "hostname" -> current?.hostName = value
                "port" -> current?.port = value.toIntOrNull()
                "user" -> current?.user = value
            }
        }
        flush()
        return result.filter { it.host.isNotEmpty() }
    }
}
